Add flash storage support to STM32WL series

fabiobaltieri opened the pull request:
This adds support for the STM32WL series flash controller, using the existing stm32l4x driver. Seems like the flash controller of the WL matches the L4 one just fine (uniform 2k pages), so it works unmodified. The name mismatch is odd but I guess there's no point making a copy just for that.

Put a patch to increase the storage partition size for nucleo_wl55jc, so that the config can be used unmodified for the settings example (it needs at least 3 flash pages, defaults to 8).

Tested with samples/subsys/settings and samples/subsys/fs/littlefs on a Seed LoRa-E5 module.

fabiobaltieri committed:
boards: nucleo_wl55jc: make the storage partition bigger

Increase the storage partition size to 16kB, 8x2kB pages. This makes the
default config usable for testing subsystems that need multiple pages.

Signed-off-by: Fabio Baltieri <fabio.baltieri@gmail.com>

fabiobaltieri committed:
drivers: flash_stm32: enable flash_stm32l4x for STM32WL series

The flash_stm32l4x driver seems to work out of the box on the WL series.

This just adds the necessary config changes to let the driver build and
run when SOC_SERIES_STM32WLX is selected.

Signed-off-by: Fabio Baltieri <fabio.baltieri@gmail.com>

erwango commented:
@fabiobaltieri would you mind rebasing ?

fabiobaltieri commented:
> @fabiobaltieri would you mind rebasing ?

Yup, rebased and pushed.

ABOSTM reviewed (changes requested):
in STM32L4 flash driver, there is management of FLASH cache (ART accelerator), but it is done 
under compilation switches
https://github.com/zephyrproject-rtos/zephyr/blob/5daf0b2fada094401bcea8ae6bb4f82c274c0ba0/drivers/flash/flash_stm32l4x.c#L76
https://github.com/zephyrproject-rtos/zephyr/blob/5daf0b2fada094401bcea8ae6bb4f82c274c0ba0/drivers/flash/flash_stm32l4x.c#L103

This ART cache management is also applicable to STM32WL,
even if STM32WL doesn't support dualbank.

fabiobaltieri commented:
> in STM32L4 flash driver, there is management of FLASH cache (ART accelerator), but it is done
> under compilation switches
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/5daf0b2fada094401bcea8ae6bb4f82c274c0ba0/drivers/flash/flash_stm32l4x.c#L76
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/5daf0b2fada094401bcea8ae6bb4f82c274c0ba0/drivers/flash/flash_stm32l4x.c#L103
> 
> This ART cache management is also applicable to STM32WL,
> even if STM32WL doesn't support dualbank.

Oh neat, ok I'm enabling it for WL as well, moved to a define on top of the file to reduce the clutter in the code. Thanks for spotting this.

fabiobaltieri commented:
Rebased, enabled dcache control when WLxx is selcted (regardless of dual bank).

erwango commented on drivers/flash/Kconfig.stm32:
```diff
@@ -29,13 +30,14 @@ config SOC_FLASH_STM32
 	  Enable STM32F0x, STM32F1x, STM32F2x, STM32F3x, STM32F4x, STM32F7x,
-	  STM32L0x, STM32L1x, STM32L4x, STM32WBx, STM32G0x, STM32G4x or STM3H7x
-	  series flash driver.
+	  STM32L0x, STM32L1x, STM32L4x, STM32WBx, STM32WLx, STM32G0x, STM32G4x
+	  or STM3H7x series flash driver.
```
I think this whole Kconfig section starts being confused and not clear enough so that users can be able to understand easily which series are supported  (and via which driver). We'll need to enhance this.
Anyway this is not blocking for current change

fabiobaltieri replied:
Yeah the list in the description does not add much value at this point.

ABOSTM commented on drivers/flash/flash_stm32l4x.c:
```diff
@@ -26,6 +26,11 @@ LOG_MODULE_REGISTER(LOG_DOMAIN);
 
+#if defined(FLASH_OPTR_DUALBANK) || defined(FLASH_OPTR_DBANK) || \
+	defined(CONFIG_SOC_SERIES_STM32WLX)
+#define CONTROL_DCACHE
```
@fabiobaltieri, I have done the job for F4 (and G4) in #36277,
you can port it to L4/WL

fabiobaltieri replied:
Oh neat, ok so the `CONTROL_DCACHE` workaround is there for the WL as well, guess it won't hurt but let me know if you want me to change the condition (or the define name). I'm adding these two then:
- add a flash cache functionality copied from https://github.com/zephyrproject-rtos/zephyr/pull/32218/commits/9ce5b887b427bfdd85df9a1eceb7fcce35fa27f4
- enabling instruction and data cache for the series copied from https://github.com/zephyrproject-rtos/zephyr/pull/36277/commits/d80835ea821569b58e60d774b883e02d5185a4c3

str4t0m replied:
I agree to the introduction of the flag, but as @ABOSTM mentioned. This errata only concerns read-while-write operations on L4, therefore it is not necessary on WL series.

fabiobaltieri replied:
Ok so I'll keep the #define but remove CONFIG_SOC_SERIES_STM32WLX from the condition.

fabiobaltieri committed:
drivers: flash: stm32f4: Flush caches after erase

This implement the same flush cache functionality already present in the
other stm32 series flash drivers, used to avoid bus errors when writing
big chunks of data to the flash.

Signed-off-by: Fabio Baltieri <fabio.baltieri@gmail.com>

fabiobaltieri committed:
soc: stm32wl: enable instruction and data cache

Enable instruction and data cache using the corresponding HAL functions.

Signed-off-by: Fabio Baltieri <fabio.baltieri@gmail.com>

fabiobaltieri commented:
Rebased and added two commits based on the review:
- add a flash cache functionality (copied from https://github.com/zephyrproject-rtos/zephyr/pull/32218/commits/9ce5b887b427bfdd85df9a1eceb7fcce35fa27f4)
- enabling instruction and data cache for the series (copied from https://github.com/zephyrproject-rtos/zephyr/pull/36277/commits/d80835ea821569b58e60d774b883e02d5185a4c3)

str4t0m reviewed (changes requested):
Thanks. This changes look fine, but the errata workaround is not needed on WL.

fabiobaltieri commented:
Dropped `CONFIG_SOC_SERIES_STM32WLX` from the list triggering the DCACHE workaround.

ABOSTM reviewed (approved):
LGTM
